### PR TITLE
LPS-169338 Created Button Action “Split Tests”

### PR DESCRIPTION
**Pull request by samuelleolopes**

### CODE REVIEW

**Comment by samuelleolopes** [comment]
## :fire: GitRay Report 
Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.
 ## Jira Tasks 

 <html><details><summary>More details about: <b>LPS-69338 Update LCS changelog</b>.</summary>
              <br />
              <table>
              <tr>
            <th>Title</th>
            <td><a target="_blank" href="https://issues.liferay.com/browse/LPS-69338">Update LCS changelog</a></td>
        </tr>
              <tr>
            <th>Type</th>
            <td>Story</td>
        </tr>
              <tr>
            <th>Priority</th>
            <td>Minor</td>
        </tr>
              <tr>
            <th>Sub-Tasks</th>
            <td>No Sub-Tasks Opened</td>
        </tr>
              <tr>
            <th>Status</th>
            <td>Closed</td>
        </tr>
              <tr>
            <th>Assignee</th>
            <td>Jeffrey Yang</td>
        </tr>
              <tr>
            <th>Reporter</th>
            <td>Jeffrey Yang</td>
        </tr>
                </table>
          </details></html> 

---

Hey, don't forget to update all related tickets with correct status, such as **In Review** :wink: 
### Git Stats 

8 files changed, 180 insertions(+), 4 deletions(-)

**Comment by samuelleolopes** [comment]
ci:test:sf

**Comment by caiosouza15** [comment]
Just starting reviewing :)

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/pages/Testflow/Subtask/SubtaskCaseResults.tsx [hunk 1]
```diff
@@ -93,6 +107,35 @@ const SubtasksCaseResults = () => {
 		return [...alerts, ...alreadyAssigned, ...statusOpen];
 	};
 
+	const onSplitSubtasks = async (
+		dispatch: Dispatch<any>,
+		items: TestraySubTaskCaseResult[],
```
Comment (rafaelfs98) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/pages/Testflow/Subtask/SubtaskCaseResults.tsx:112]:
```suggestion
		subTaskCaseResults: TestraySubTaskCaseResult[],
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/pages/Testflow/Subtask/SubtaskCaseResults.tsx [hunk 2]
```diff
@@ -93,6 +107,35 @@ const SubtasksCaseResults = () => {
...
+		mutate: KeyedMutator<any>,
+		subTaskId: string,
+		subtasksCaseResults: TestraySubTaskCaseResult[]
```
Comment (rafaelfs98) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/pages/Testflow/Subtask/SubtaskCaseResults.tsx:115]:
```suggestion
		selectedCaseResults: TestraySubTaskCaseResult[]
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/pages/Testflow/Subtask/SubtaskCaseResults.tsx [hunk 3]
```diff
@@ -93,6 +107,35 @@ const SubtasksCaseResults = () => {
...
+	) => {
+		await testraySubTaskImpl.splitToSubtask(
+			items,
```
Comment (rafaelfs98) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/pages/Testflow/Subtask/SubtaskCaseResults.tsx:118]: allCaseResults

Comment (rafaelfs98) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/pages/Testflow/Subtask/SubtaskCaseResults.tsx:118]:
```suggestion
			allCaseResults,
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/pages/Testflow/Subtask/SubtaskCaseResults.tsx [hunk 4]
```diff
@@ -204,6 +251,15 @@ const SubtasksCaseResults = () => {
 							})
 						}
 						isVisible={!!selectedRows.length}
+						onSubmit={() =>
+							onSplitSubtasks(
+								dispatch,
+								items,
+								mutate,
+								subtaskId as string,
+								selectedResult
```
Comment (rafaelfs98) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/pages/Testflow/Subtask/SubtaskCaseResults.tsx:260]:
```suggestion
								selectedCaseResults
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestraySubtask.ts [hunk 5]
```diff
@@ -217,6 +222,66 @@ class TestraySubtaskImpl extends Rest<SubtaskForm, TestraySubTask> {
...
+
+	public async splitToSubtask(
+		items: TestraySubTaskCaseResult[],
+		subTaskCaseResults: TestraySubTaskCaseResult[],
+		subTaskId: string,
+		subTasks: TestraySubTask[]
+	) {
+		const listSubTask = subTaskCaseResults.map(({subTask}) => subTask);
+
+		const caseResultsUpdate = subTaskCaseResults.map(({id}) => id);
```
Comment (rafaelfs98) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestraySubtask.ts:234]:
```suggestion
		const caseResultIds = subTaskCaseResults.map(({id}) => id);
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestraySubtask.ts [hunk 6]
```diff
@@ -217,6 +222,66 @@ class TestraySubtaskImpl extends Rest<SubtaskForm, TestraySubTask> {
...
+			number: newNumSub,
+			score: sumPriority,
+			splitFromSubtaskId: listSubTask[0]?.id,
+			taskId: listSubTask[0]?.r_taskToSubtasks_c_task.id,
+			userId: listSubTask[0]?.r_userToSubtasks_user.id,
+		};
+
+		const newSubtask = await super.create(createSubtask as SubtaskForm);
+
+		for (const caseResult of caseResultsUpdate) {
```
Comment (rafaelfs98) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestraySubtask.ts:275]:
```suggestion
		for (const caseResult of caseResultIds) {
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestraySubtask.ts [hunk 7]
```diff
@@ -217,6 +222,66 @@ class TestraySubtaskImpl extends Rest<SubtaskForm, TestraySubTask> {
...
+		items: TestraySubTaskCaseResult[],
+		subTaskCaseResults: TestraySubTaskCaseResult[],
+		subTaskId: string,
+		subTasks: TestraySubTask[]
+	) {
+		const listSubTask = subTaskCaseResults.map(({subTask}) => subTask);
+
+		const caseResultsUpdate = subTaskCaseResults.map(({id}) => id);
+
+		const contPrioritySelect = subTaskCaseResults.map(
```
Comment (rafaelfs98) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestraySubtask.ts:236]:
```suggestion
		const casePrioritySelected = subTaskCaseResults.map(
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestraySubtask.ts [hunk 8]
```diff
@@ -217,6 +222,66 @@ class TestraySubtaskImpl extends Rest<SubtaskForm, TestraySubTask> {
...
+			({caseResult}) => caseResult?.case?.priority ?? 0
+		);
+
+		const sumPriority = contPrioritySelect.reduce(
```
Comment (rafaelfs98) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestraySubtask.ts:240]:
```suggestion
		const sumPriority = casePrioritySelected.reduce(
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestraySubtask.ts [hunk 9]
```diff
@@ -217,6 +222,66 @@ class TestraySubtaskImpl extends Rest<SubtaskForm, TestraySubTask> {
...
+			(prev, next) => prev + next
+		);
+
+		const numSub = subTasks.map(({number}) => number);
```
Comment (rafaelfs98) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestraySubtask.ts:244]:
```suggestion
		const subtaskNumber = subTasks.map(({number}) => number);
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestraySubtask.ts [hunk 10]
```diff
@@ -217,6 +222,66 @@ class TestraySubtaskImpl extends Rest<SubtaskForm, TestraySubTask> {
...
+
+		const numMax = numSub.reduce((prev, next) => {
```
Comment (rafaelfs98) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestraySubtask.ts:246]:
```suggestion
		const subtaskNumberMax = subtaskNumber.reduce((prev, next) => {
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestraySubtask.ts [hunk 11]
```diff
@@ -217,6 +222,66 @@ class TestraySubtaskImpl extends Rest<SubtaskForm, TestraySubTask> {
...
+			return Math.max(prev, next);
+		});
+
+		const newNumSub = numMax + 1;
```
Comment (rafaelfs98) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestraySubtask.ts:250]:
```suggestion
		const newSubtaskNumber = numMax + 1;
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestraySubtask.ts [hunk 12]
```diff
@@ -217,6 +222,66 @@ class TestraySubtaskImpl extends Rest<SubtaskForm, TestraySubTask> {
 			score: sumScore,
 		});
 	}
+
+	public async splitToSubtask(
+		items: TestraySubTaskCaseResult[],
```
Comment (rafaelfs98) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestraySubtask.ts:227]:
```suggestion
		allSubtaskCaseResults: TestraySubTaskCaseResult[],
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestraySubtask.ts [hunk 13]
```diff
@@ -217,6 +222,66 @@ class TestraySubtaskImpl extends Rest<SubtaskForm, TestraySubTask> {
...
+
+		const numSub = subTasks.map(({number}) => number);
+
+		const numMax = numSub.reduce((prev, next) => {
+			return Math.max(prev, next);
+		});
+
+		const newNumSub = numMax + 1;
+
+		const numSplitUpdate = items
```
Comment (rafaelfs98) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestraySubtask.ts:252]:
```suggestion
		const numSplitUpdate = allSubtaskCaseResults
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestraySubtask.ts [hunk 14]
```diff
@@ -217,6 +222,66 @@ class TestraySubtaskImpl extends Rest<SubtaskForm, TestraySubTask> {
 			score: sumScore,
 		});
 	}
+
+	public async splitToSubtask(
+		items: TestraySubTaskCaseResult[],
+		subTaskCaseResults: TestraySubTaskCaseResult[],
```
Comment (rafaelfs98) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestraySubtask.ts:228]:
```suggestion
		selectedSubTaskCaseResults: TestraySubTaskCaseResult[],
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestraySubtask.ts [hunk 15]
```diff
@@ -217,6 +222,66 @@ class TestraySubtaskImpl extends Rest<SubtaskForm, TestraySubTask> {
...
+
+		const numMax = numSub.reduce((prev, next) => {
+			return Math.max(prev, next);
+		});
+
+		const newNumSub = numMax + 1;
+
+		const numSplitUpdate = items
+			.filter(
+				(item) => !subTaskCaseResults.find((select) => item === select)
```
Comment (rafaelfs98) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestraySubtask.ts:254]:
```suggestion
				(subtaskCaseResult) => !selectedSubTaskCaseResults.find((selected) => subtaskCaseResult === selected)
```

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestraySubtask.ts [hunk 16]
```diff
@@ -217,6 +222,66 @@ class TestraySubtaskImpl extends Rest<SubtaskForm, TestraySubTask> {
...
+
+		const sumNewPriority = numSplitUpdate.reduce(
+			(prev, next) => prev + next
+		);
+
+		const createSubtask = {
+			dueStatus: listSubTask[0]?.dueStatus.key,
+			errors: subTaskCaseResults[0]?.caseResult?.errors || ' ',
+			name: `ST-${newNumSub}`,
+			number: newNumSub,
```
Comment (rafaelfs98) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/services/rest/TestraySubtask.ts:266]:
```suggestion
			number: newSubtaskNumber
```

**Review by rafaelfs98 — COMMENTED** [review]
the code looks good, congrats man! fire

I left a few suggestions for you

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/pages/Testflow/Subtask/SubtaskCaseResults.tsx [hunk 17]
```diff
@@ -93,6 +107,35 @@ const SubtasksCaseResults = () => {
 		return [...alerts, ...alreadyAssigned, ...statusOpen];
 	};
 
+	const onSplitSubtasks = async (
+		dispatch: Dispatch<any>,
+		allCaseResults: TestraySubTaskCaseResult[],
+		mutate: KeyedMutator<any>,
```
Comment (welitonpe) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/pages/Testflow/Subtask/SubtaskCaseResults.tsx:113]: If you can change the any of Dispatch and the KeyedMutator.

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/pages/Testflow/Subtask/SubtaskOutlet.tsx [hunk 18]
```diff
@@ -60,12 +68,26 @@ const SubtaskOutlet = () => {
 		(response) => testraySubtaskIssuesImpl.transformDataFromList(response)
 	);
 
+	const {data: testraySubtaskToSplit} = useFetch<APIResponse<TestraySubTask>>(
+		`${testraySubTaskImpl.resource}&filter=${searchUtil.eq(
+			'r_splitFromTestraySubtask_c_subtaskId',
```
Comment (welitonpe) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/extra/remote-app/src/pages/Testflow/Subtask/SubtaskOutlet.tsx:73]: you're using the name of relationship, you can get a short name trough transform data, but you need to implement this on TestraySubtask.ts file.

File: modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/src/main/resources/site-initializer/object-definitions/testray-subtask.json [hunk 19]
```diff
@@ -48,6 +48,17 @@
...
 			"type": "String"
 		},
+		{
+			"DBType": "Integer",
+			"indexed": false,
+			"indexedAsKeyword": false,
+			"label": {
+				"en_US": "Number"
+			},
+			"name": "number",
```
Comment (welitonpe) [inline modules/apps/site-initializer/site-initializer-testray/site-initializer-testray/src/main/resources/site-initializer/object-definitions/testray-subtask.json:58]: you can change name of this field of object.

**Review by welitonpe — COMMENTED** [review]
Need some adjust.

**Comment by samuelleolopes** [comment]
ci:test:sf

**Comment by kevenleone** [comment]
@samuelleolopes can you rebase with master? because we have conflicting file changes

**Comment by kevenleone** [comment]
@samuelleolopes your last commit is:
LPS-69338 Created Button Action “Split Tests”

Is missing `1` LPS-169338 Created Button Action “Split Tests” in LPS

**Review by welitonpe — APPROVED** [review]
Good Job :fire:

**Comment by welitonpe** [comment]
ci:test:sf

**Comment by samuelleolopes** [comment]
ci:test:sf

**Review by caiosouza15 — APPROVED** [review]
LGTM :fire:

**Comment by kevenleone** [comment]
Thank you guys, resent: https://github.com/pt-liferay-solutions/liferay-portal/pull/2911

**Pull request closed without merging**